[IDE] Add full documentation to code completion result

a7medev opened the pull request:
## Full documentation retrieval

Retrieve and return full documentation in code completion results and expose it in:
- SourceKit plugin through `swiftide_completion_item_get_doc_full`
- Code completion request (used by `sourcekitd-test` though not used in SourceKit LSP) in the `key.doc.full_as_xml` key.

## Round-trippable declarations

We need the `Decl` instance to retrieve the full documentation comment for a code completion item.
To avoid having to fetch and cache full documentation comments for all cached code completion items (for performance), we'll need a way to retrieve the `Decl` associated with a code completion item after retrieving it from the cache.

Since `Decl` instances aren't cacheable, we work around that by retrieving the `Decl` instance using the primary associated USR through:
1. Turning the USR into a mangled name.
2. Demangling the mangled name to get a demangling AST.
3. We extract the declaration name from the demangling AST (using the newly added `DeclNameExtractor`).
4. We retrieve the `DeclContext` using the existing `ASTBuilder::findDeclContext`.
5. We perform a lookup for declarations with the provided name in the `DeclContext`.
6. We search the lookup results for a declaration that has a matching USR, if found we return it, otherwise we return `nullptr`.

### Testing strategy

Since there are a lot of cases for declarations and this is mainly meant for use within code completion to retrieve the full documentation through the associated declaration, we're making use of existing code completion tests by:
1. Adding a "verify USR to Decl" option to completion requests which when enabled tries to reconstruct all code completion results' associated declarations using their Swift USR. This option is only meant for use within test cases and is thus disabled by default on the request-level.
2. Enabling the "verify USR to Decl" option in `sourcekitd-test`, `complete-test`, and `swift-ide-test` by default to verify reconstruction for all existing tests.

a7medev committed:
[IDE] Add full documentation to code completion result

a7medev committed:
[IDE] Rename complete_docbrief to complete_doc

a7medev committed:
[Demangle] Implement Swift USR to Decl conversion

a7medev committed:
[Demangle] Separate DeclNameExtractor & Clang USR to Decl

a7medev committed:
[IDE] Export swiftide_completion_item_get_doc_full

a7medev committed:
[USR] Allow generating Swift USR for any Clang decl

a7medev committed:
[Demangle] Optimize USR to Clang Decl performance

a7medev committed:
[Demangle] Find extension by underlying module for Clang decls

a7medev committed:
[Demangle] Rename getDeclFromUSR to getDeclForUSR for consistency

a7medev committed:
[Demangle] Fix issues with Clang USR to Decl reconstruction

a7medev committed:
[IDE] Add AssociatedDecl to CodeCompletionResult

a7medev committed:
[IDE] Pass AssociatedDecl to CodeCompletionResult on creation

a7medev committed:
[IDE] Fetch declaration full documentation on-demand

a7medev committed:
[IDE] Store Swift USR in ContextFreeCodeCompletionResult to find AssociatedDecl

a7medev committed:
[IDE] Enable full doc option in complete tests that assert on doc

a7medev committed:
[Demangle] Remove unused lookup DC param to getDeclForUSR

a7medev committed:
[Demangle] Move move switch' default at the bottom

a7medev committed:
[IDE] Verify USR to Decl on code completion

a7medev committed:
[IDE] Add verify USR to Decl completion option

a7medev committed:
[IDE] Don't pass ASTContext directly to verifyUSRToDeclReconstruction

a7medev committed:
[Demangle] Compare extension module against underlying module in findDeclContext

a7medev committed:
[Test] Default to verify USR to Decl in tests

a7medev committed:
[Test] Don't check equality of decls in verifyUSRToDeclReconstruction

The declarations may not be identical (have the same pointer) but if the reconstructed declaration is not null, it's guaranteed to have the same USR as the original declaration so they refer to the same logical declaration.

a7medev committed:
[Demangle] Fix DeclName extraction for raw identifiers

a7medev committed:
[Demangle] Extract identifier from PrivateDeclName & RelatedEntityDeclName nodes in DeclNameExtractor

a7medev committed:
[ASTContext] Remove module from NameToModules on load module with same name

a7medev committed:
[Demangle] Remove unqualified lookup in ASTBuilder::findDecl

a7medev committed:
[Demangle] Lookup Decl with privateDiscriminator in ASTBuilder::findDecl

a7medev committed:
[Test] Fix invalid redeclarations in completion tests & disable verify USR to Decl for protocol where

a7medev committed:
[IDE] Fetch completion result associated decl on-demand on printFullDocComment

a7medev committed:
[IDE] Pass mutable C-string to handler in swiftide_completion_item_get_doc_full_copy

hamishknight commented on lib/AST/ASTDemangler.cpp:
```diff
@@ -1363,19 +1487,24 @@ ASTBuilder::findDeclContext(NodePointer node) {
 
-      for (auto *module : potentialModules)
-        if (auto typeDecl = findTypeDecl(module, getIdentifier(name),
+      // FIXME: Looping over potential modules in reverse is a hack to check the
+      //        current main module before checking other ModuleDecls in the
+      //        NameToModules map to avoid an issue where a DeclContext exists
+      //        but doesn't include all desired declarations.
+      for (auto *module : reverse(potentialModules))
```
As discussed, I think we should only be considering loaded modules here. Actually thinking about it further, I suspect we could probably just fix this in `addLoadedModule`, if there's already an existing module in `LoadedModules` we should remove it from `NameToModules` too.

a7medev replied:
I updated it and it works fine now, thanks for the suggestion 🙏🏼

hamishknight commented on lib/AST/USRGeneration.cpp:
```diff
@@ -260,6 +249,28 @@ swift::USRGenerationRequest::evaluate(Evaluator &evaluator, const ValueDecl *D,
+  if (isa<ModuleDecl>(D))
+    return std::string(); // Ignore.
+
+  if (!options.useSwiftUSR) {
+    if (auto USR = generateClangUSR(D, options))
+      return USR.value();
+  }
```
One thing we could do to make this more efficient would be to split the `USRGenerationRequest` request into a request for the Clang USR that uses split caching, and a separate request for the Swift USR. That way we wouldn't need to compute the Swift USR twice.

a7medev replied:
I understand the problem but I want to make sure I understand the proposed solution correctly.

Swift USRs aren't affected by the `useSwiftUSR` option and thus will get calculated and cached twice, so we want to optimize that by having different requests for Clang and Swift each with its own caching and the `USRGenerationRequest` would just act as an umbrella that forwards requests to these two.

After performing this modification, should we cache `USRGenerationRequest` as well? Or should it not be cached since it's going to call into requests that have caching anyway? If we cache `USRGenerationRequest` results for Swift decls for example we will end up storing the same USR three times (once in `SwiftUSRGenerationRequest`'s cache, once in `USRGenerationRequest` for `useSwiftUSR == false`, and once again in `USRGenerationRequest` for `useSwiftUSR == true`.

What do you think?

hamishknight replied:
> Swift USRs aren't affected by the `useSwiftUSR` option and thus will get calculated and cached twice, so we want to optimize that by having different requests for Clang and Swift each with its own caching and the `USRGenerationRequest` would just act as an umbrella that forwards requests to these two.

Yes that's correct

> After performing this modification, should we cache `USRGenerationRequest` as well?

No, that should become uncached (`RequestFlags::Uncached`)

> I'm not sure what you mean by "split caching" though.

See e.g `OpaqueResultTypeRequest`, we mark the request as `RequestFlags::SeparatelyCached | RequestFlags::SplitCached`, which lets us store a bit on the Decl's `LazySemanticInfo` to represent the case where know that a negative result was produced in `cacheResult`. We'd want to do this for the Clang USR request for cases where the decl is not representable by a Clang USR (mainly this is for handling the `shouldUseObjCUSR` case where we use a Clang USR for a Swift decl if it can be exposed to Obj-C).

a7medev replied:
Done, please check 🙏🏼

hamishknight reviewed (commented):
Wonderful, thank you!

a7medev committed:
[IDE] Verify decls match in verify USR to Decl reconstruction

a7medev committed:
[Demangle] Enhance coding style in ASTDemangler & DeclNameExtractor

a7medev committed:
[Demangle] Generalize ASTDemangler::findDecl to take a predicate function instead of fixed USR matching

a7medev commented on lib/AST/DeclNameExtractor.cpp:
```diff
@@ -0,0 +1,238 @@
+    return;
+  }
+  
+  // More than one parameter are present
+  while (Params && Params->getFirstChild() &&
+         Params->getFirstChild()->getKind() != Node::Kind::TupleElement) {
+    Params = Params->getFirstChild();
+  }
```
This was done to resemble the logic in [swift_getFunctionFullNameFromMangledName of Casting.cpp](https://github.com/swiftlang/swift/blob/5d98ca63c738cfa36b122909aed2c26d091ebdff/stdlib/public/runtime/Casting.cpp#L349-L354), I wasn't able to reproduce such case though.

a7medev commented on lib/IDE/CodeCompletionResultBuilder.cpp:
```diff
@@ -80,15 +82,66 @@ copyAssociatedUSRs(llvm::BumpPtrAllocator &Allocator, const Decl *D) {
+  if (ide::printValueDeclSwiftUSR(VD, OS)) {
+    llvm::errs() << "Declaration is:\n";
+    VD->dump(llvm::errs());
+    assert(false && "Declaration should have a Swift USR");
```
I couldn't find a function/macro named `ABORT`, the closest I know of is `llvm_unreacheable` but I think of it as representing a case that isn't handled or shouldn't be reached if the implementation is correct. While in this case what we need is an unconditional assertion failure (similar to `assertionFailure` in Swift) and I see the `assert(false && "message")` is relatively common across the codebase for such a use case.

hamishknight replied:
It's defined here https://github.com/swiftlang/swift/blob/bb4f61e97bd413a2f695c2e4308383383fed7012/include/swift/Basic/Assertions.h#L200-L215

hamishknight replied:
Actually I would recommend switching to the lambda version of `ABORT` and putting the logging logic in there

hamishknight commented on tools/SourceKit/tools/sourcekitd/bin/InProc/CodeCompletionSwiftInterop.h:
```diff
@@ -272,6 +272,9 @@ SWIFTIDE_PUBLIC void
 swiftide_completion_request_set_add_call_with_no_default_args(
     swiftide_completion_request_t, bool);
 
+SWIFTIDE_PUBLIC void swiftide_completion_request_set_verify_usr_to_decl(
+    swiftide_completion_request_t, bool);
```
Given it's only used for testing, I don't think there's much value in exposing it if SourceKit-LSP isn't going to use it (we could potentially enable it in its test suite, but I'm not convinced that will give us too much more coverage than what we're already getting from the Swift test suite)

a7medev replied:
Makes sense, I'll remove it.

a7medev commented on tools/SourceKit/lib/SwiftLang/SwiftCompletion.cpp:
```diff
@@ -684,14 +700,17 @@ bool CodeCompletion::SessionCacheMap::remove(StringRef name, unsigned offset) {
 namespace {
 class SwiftGroupedCodeCompletionConsumer : public CodeCompletionView::Walker {
   GroupedCodeCompletionConsumer &consumer;
+  bool includeFullDocumentation;
```
Makes sense, it was also used in `complete-test` but it's still testing the `complete.open` flow so I removed it.

a7medev committed:
[USR] Split USRGenerationRequest into SwiftUSRGenerationRequest & ClangUSRGenerationRequest

With the introduction of useSwiftUSR option to USRGenerationRequest, Swift declaration USRs could now get cached twice even though the useSwiftUSR option doesn't affect Swift declarations.
We turned USRGenerationRequest into an uncached request that forwards requests to SwiftUSRGenerationRequest and ClangUSRGenerationRequest which cache results.
ClangUSRGenerationRequest uses split caching to avoid filling the cache with a lot of null results.

a7medev committed:
[IDE] Switch assert false to ABORT in verifyUSRToDeclReconstruction

a7medev committed:
[IDE] Apply clang-format to full documentation changes

a7medev committed:
[IDE] Update code complete and documentation tests

a7medev committed:
[IDE] NFC: Add doc comment for on-demand CodeCompletionResult::getAssociatedDecl

a7medev commented on include/swift/AST/DeclNameExtractor.h:
```diff
@@ -0,0 +1,58 @@
+                         Identifier &privateDiscriminator);
+
+} // end namespace swift
+} // end namespace Demangle
```
Good catch, fixed ✅

a7medev commented on lib/AST/ASTDemangler.cpp:
```diff
@@ -1584,28 +1717,3 @@ GenericTypeDecl *ASTBuilder::findForeignTypeDecl(StringRef name,
   return consumer.Result;
 }
-
-Identifier ASTBuilder::getIdentifier(StringRef name) {
```
That's a good idea, done ✅

a7medev commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -764,13 +764,18 @@ struct USRGenerationOptions {
+  bool useSwiftUSR;
+
   friend llvm::hash_code hash_value(const USRGenerationOptions &options) {
     return llvm::hash_value(options.distinguishSynthesizedDecls);
```
Good catch, done ✅

a7medev commented on lib/AST/ASTDemangler.cpp:
```diff
@@ -75,6 +95,112 @@ TypeDecl *swift::Demangle::getTypeDeclForUSR(ASTContext &ctx,
+  if (!contextNode)
+    return nullptr;
+  
+  auto hasMatchingUSR = [usr](const ValueDecl *VD) {
```
Done ✅

a7medev commented on lib/IDE/CodeCompletionResultBuilder.cpp:
```diff
@@ -186,9 +241,16 @@ CodeCompletionResult *CodeCompletionResultBuilder::takeResult() {
+    if (verifyUSRToDecl) {
+      auto &Ctx = DC->getASTContext();
+      verifyUSRToDeclReconstruction(Ctx, AssociatedDecl);
+    }
```
Yeah I missed this one 😅, removed it now.

a7medev commented on lib/AST/DeclNameExtractor.cpp:
```diff
@@ -0,0 +1,238 @@
+    return extractFunctionLikeName(node, name, privateDiscriminator);
+    
+  case Node::Kind::Subscript:
+    return extractFunctionLikeName(node, name, privateDiscriminator);
```
Done. ✅

a7medev commented on lib/AST/DeclNameExtractor.cpp:
```diff
@@ -0,0 +1,238 @@
+  }
+}
+
+void DeclNameExtractor::extractArgLabelsFromType(Node *Type, SmallVectorImpl<Identifier> &ArgLabels) {
```
Done ✅

a7medev commented on test/IDE/complete_crashes.swift:
```diff
@@ -1,7 +1,7 @@
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-verify-usr-to-decl=false -code-completion-token=BAD_MEMBERS_1 | %FileCheck %s -check-prefix=BAD_MEMBERS_1
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-verify-usr-to-decl=false -code-completion-token=BAD_MEMBERS_2 | %FileCheck %s -check-prefix=BAD_MEMBERS_2
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-verify-usr-to-decl=false -code-completion-token=CLOSURE_CALLED_IN_PLACE_1 | %FileCheck %s -check-prefix=WITH_GLOBAL_INT
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -code-completion-verify-usr-to-decl=false -code-completion-token=RDAR_28991372 | %FileCheck %s -check-prefix=RDAR_28991372
```
As discussed before I was disabling these due to the name conflict between `class C` and `class C<T>` causing USR to Decl to find the wrong decl. I've fixed the conflict and enabled it in this test.

a7medev commented on test/SourceKit/CodeComplete/complete_member.swift:
```diff
@@ -39,9 +39,60 @@ func testOverrideUSR() {
-// RUN: %sourcekitd-test -req=complete -pos=15:5 %s -- %s > %t.response
-// RUN: %diff -u %s.response %t.response
-//
+// RUN: %sourcekitd-test -req=complete -req-opts=includefulldocumentation=1 -pos=15:5 %s -- %s | %FileCheck %s -check-prefix=CHECK-NORMAL
```
I've reverted the change to remove `complete_members.swift.response` and used `%sed_clean` to clean up the absolute path so I think that should work fine. Let me know if you'd still prefer `split-file` though.

a7medev commented on include/swift/IDE/CodeCompletionResult.h:
```diff
@@ -716,6 +729,8 @@ class CodeCompletionResult {
     }
   }
 
+  const Decl *getAssociatedDecl();
```
Done ✅

a7medev commented on include/swift/IDE/CodeCompletionResultSink.h:
```diff
@@ -40,6 +40,12 @@ struct CodeCompletionResultSink {
+  bool verifyUSRToDecl = false;
+
+  /// Whether to include full documentation in all completion items.
+  bool includeFullDocumentation = false;
```
It's not needed anymore now that I've removed its use from the `complete.open`/`complete.update` requests and remove the option from `complete-test` and `swift-ide-test` so I removed it.

a7medev commented:
@hamishknight Thanks for the review. I've addressed all comments, please let me know if there's anything else that needs to be changed 🙏🏼

hamishknight reviewed (commented):
Thank you! Looks great, just a couple more comments

a7medev commented on tools/SourceKit/tools/sourcekitd/bin/InProc/CodeCompletionSwiftInterop.cpp:
```diff
@@ -661,6 +663,23 @@ void swiftide_completion_item_get_doc_brief(
+  if (buffer.empty()) {
+    return handler(nullptr);
+  }
+  handler(strdup(buffer.c_str()));
```
I think it was a mistake due to my current git workflow 😅, thanks for pointing it out

a7medev replied:
I think it was a mistake due to my current git workflow 😅, thanks for pointing it out

a7medev committed:
[IDE] Use response scratch in swiftide_completion_item_get_doc_full

a7medev committed:
[USR] Remove unneeded code in USR to Decl conversion

a7medev committed:
[Test] Don't print brief comment in diagnostics in swift-ide-test

rintaro commented on tools/SourceKit/lib/SwiftLang/CodeCompletionOrganizer.h:
```diff
@@ -46,6 +46,8 @@ struct Options {
   bool includeObjectLiterals = true;
   bool addCallWithNoDefaultArgs = true;
+  bool verifyUSRToDecl = false;
+  bool includeFullDocumentation = false;
```
I don't see anything is actually using this flag now. I believe this is a leftover from the previous implementation. Could you remove this?

a7medev replied:
It's used for testing full documentation retrieval using `sourcekitd-test`.

rintaro replied:
Oh I see, sorry I misunderstood. That being said, I don't think we should have this option since we don't want anything would use it as SourceKit API. For testing the full documentation rendering, I believe `swift-ide-test` is enough. I'd prefer not to have sourcekitd-test testing the `doc.full_as_xml`. @hamishknight @bnbarham wdyt?

bnbarham replied:
Yeah, I think swift-ide-test + sourcekit-lsp testing is enough here. The sourcekitd-test as is doesn't really test anything that swift-ide-test isn't already.

rintaro commented on tools/SourceKit/include/SourceKit/Core/LangSupport.h:
```diff
@@ -90,6 +90,7 @@ struct CodeCompletionInfo {
   StringRef TypeName;
   StringRef ModuleName;
   StringRef DocBrief;
+  StringRef DocFull;
```
Ditto

rintaro commented on tools/SourceKit/lib/SwiftLang/SwiftCompletion.cpp:
```diff
@@ -722,12 +741,15 @@ static void translateCodeCompletionOptions(OptionsDictionary &from,
   static UIdent KeyAddInitsToTopLevel("key.codecomplete.addinitstotoplevel");
   static UIdent KeyFuzzyMatching("key.codecomplete.fuzzymatching");
+  static UIdent KeyIncludeFullDocumentation(
+      "key.codecomplete.includefulldocumentation");
```
Ditto

rintaro commented on tools/SourceKit/tools/sourcekitd/lib/API/CodeCompletionResultsArray.cpp:
```diff
@@ -24,8 +24,8 @@ using namespace sourcekitd;
                       std::optional<StringRef>, std::optional<StringRef>,
-                      std::optional<StringRef>, UIdent, UIdent, uint8_t,
-                      uint8_t>
+                      std::optional<StringRef>, std::optional<StringRef>,
```
Ditto

bnbarham commented:
I don't have any more comments on top of what @hamishknight and @rintaro already have. But I just wanted to say that this is awesome work @a7medev!

a7medev commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -764,13 +764,21 @@ struct USRGenerationOptions {
+
   friend llvm::hash_code hash_value(const USRGenerationOptions &options) {
-    return llvm::hash_value(options.distinguishSynthesizedDecls);
+    return llvm::hash_combine(
+        llvm::hash_value(options.distinguishSynthesizedDecls),
+        llvm::hash_value(options.useSwiftUSR));
   }
 
   friend bool operator==(const USRGenerationOptions &lhs,
                          const USRGenerationOptions &rhs) {
-    return lhs.distinguishSynthesizedDecls == rhs.distinguishSynthesizedDecls;
+    return lhs.distinguishSynthesizedDecls == rhs.distinguishSynthesizedDecls &&
+           lhs.useSwiftUSR == rhs.useSwiftUSR;
   }
 
   friend bool operator!=(const USRGenerationOptions &lhs,
```
We still need them in `SimpleRequest`'s `==`, `!=` operators and `hash_value` implementations even if the request is uncached.

a7medev committed:
[USR] Move distinguishSynthesizedDecls check from ClangUSRGenerationRequest to USRGenerationRequest

a7medev committed:
[Demangle] Simplify DeclNameExtractor::extractArgLabelsFromType multiple parameters case

a7medev committed:
[IDE] Remove includeFullDocumentation option in code complete request

a7medev committed:
[Test] Check full doc in completion with swift-ide-test

a7medev commented on tools/SourceKit/tools/sourcekitd-test/CMakeLists.txt:
```diff
@@ -12,11 +12,11 @@ target_link_libraries(sourcekitd-test PRIVATE
+#if(SWIFT_SOURCEKIT_USE_INPROC_LIBRARY)
   target_link_libraries(sourcekitd-test PRIVATE sourcekitdInProc)
-else()
-  target_link_libraries(sourcekitd-test PRIVATE sourcekitd)
-endif()
+#else()
+#  target_link_libraries(sourcekitd-test PRIVATE sourcekitd)
+#endif()
```
I'm mostly using incremental builds with `ninja`, is there a way to pass this to `ninja` as well?

a7medev commented on test/IDE/complete_where_clause.swift:
```diff
@@ -43,6 +41,12 @@
 
+// FIXME: We disable USR to Decl verification in the following 2 tests since for
+//        some reason `ASTBuilder::createTypeDecl` fails to find the DeclContext
+//        (the protocol) containing the associated type.
```
As discussed, this is happening due to extension binding triggering a module lookup before all declarations have been parsed causing the creation of an incomplete `SourceLookupCache` for the main module which doesn't get updated when the rest of the declarations are added later using `SourceFile::addTopLevelDecl`.

I've updated the comment as agreed.

a7medev committed:
NFC: Apply clang-format to full documentation in completion changes

a7medev commented:
Thanks for the thorough review 🙏🏼 
@hamishknight @rintaro I've addressed the comments you added, please recheck.

rintaro commented on test/SourceKit/CodeComplete/complete_doc_1.swift:
You don't need to update `SourceKit` tests. Apologies if I wasn't clear, but I meant `test/IDE/complete_*` tests should be enough for testing full documentation rendering. If you don't think please add new test cases to `test/IDE`.

`test/SourceKit` directory is for testing sourcekitd specific functionalities, like response format, etc.

hamishknight reviewed (approved):
Thank you!

hamishknight commented on test/SourceKit/CodeComplete/complete_doc_1.swift:
```diff
@@ -0,0 +1,34 @@
+}
+
+// All in main module.
+// RUN: %target-swift-ide-test -code-completion -source-filename %s -module-name DocBriefTest -code-completion-token=DOC -code-completion-comments=true | %FileCheck %s -check-prefix=CHECK
```
Oh heh, raced with Rintaro's comment. Let's keep the SourceKit brief comment tests and add new ones in `test/IDE` for full + brief

a7medev committed:
[IDE] Remove unused DeclNameExtractor::extractTextName & SourceKit Completion::printFullDocComment

a7medev committed:
[Test] Revert changes to SourceKit documentation tests

a7medev committed:
[Test] Add IDE tests for full and brief documentation

a7medev commented:
@rintaro @hamishknight I've reverted changes to SourceKit tests (except for invalid redeclaration fixes) and added some tests under `test/IDE`. Can you please recheck? 🙏🏼

rintaro commented:
@swift-ci Please test

a7medev commented:
Some tests are failing due to failures in USR to Decl verification. I suspect this is the same issue we had with extension binding causing the creation of an incomplete `SourceLookupCache` for the module. 3 tests may be related to extension binding (though I haven't confirmed), 1 doesn't.

I'm looking into it.

a7medev committed:
[Demangle] Demangle operator function name in DeclNameExtractor

a7medev committed:
[Test] Disable USR to Decl verification in crashers with invalid nameless extension

The USR for declarations within the extension with no type name have a USR that looks like a top-level declaration causing the lookup within swift::Demangle::getDeclForUSR to fail.

a7medev commented:
Failing tests can be categorized into 3 buckets:
1. `IDE/crashers_fixed/003-swift-typebase-getcanonicaltype.swift`, `IDE/crashers_fixed/006-swift-declcontext-getprotocolself.swift`, and `IDE/crashers_fixed/064-swift-genericfunctiontype-get.swift`: these have decls within an extension with no type name (`extension {`), causing these decls to have a USR that does looks like a top-level decl's USR, which causes the lookup inside `Demangle::getDeclForUSR` to fail. There's not much, in the scope of this PR, that we can do here, so I disabled USR to Decl verification for these tests.
2. `IDE/crashers_fixed/093-swift-valuedecl-getformalaccessscope.swift`: the test has an (invalid) function operator decl `func <()` (created by parser after recovery), we didn't handle operators in `DeclNameExtractor` since their `FuncDecl`s weren't passed to `CodeCompletionResultBuilder::setAssociatedDecl` but rather their `OperatorDecl`s which is not a `ValueDecl`. I added support for operator name extraction in `DeclNameExtractor` to fix this and also for future changes where we pass the operator's `FuncDecl` rather than `OperatorDecl` to `CodeCompletionResultBuilder` (this will allow us to show brief & full documentation for operators which is broken now).
3. `IDE/complete_from_cocoa.swift`, `IDE/complete_from_cocoa_2.swift`, and `IDE/complete_sdk_platform.swift`: USR to Decl reconstruction fails since:
    1. The `EnumDecl` for an Objective-C enum (`AppKit.NSLineBreakMode`) is added to completion results and USR to Decl verification kicks off (with Swift USR: `s:So15NSLineBreakModeV`).
    2. When we try to find the `NSLineBreakMode` type using `createTypeDecl`, `ClangImporter::lookupTypeDecl` does find the `clang::EnumDecl` but when it imports it using `ClangImporter::Implementation::importDecl` it returns a `TypeAliasDecl` with the name `LineBreakMode` instead (a type-alias to the real enum).
    3. Verification fails since the original `EnumDecl` != the reconstructed `TypeAliasDecl` crashing the tests.

    I'm still not sure why this behavior happens, and whether we should fix it in this PR or leave it for now.

a7medev commented on lib/AST/DeclNameExtractor.cpp:
```diff
@@ -86,10 +86,11 @@ bool DeclNameExtractor::extractFunctionLikeName(
+    if (extractIdentifierName(node, name, privateDiscriminator)) {
+      BaseName = name.getBaseName();
+    } else {
+      BaseName = extractOperatorName(node);
```
It will return `false` in the condition below:
```cpp
  if (BaseName.empty())
    return false;
```

a7medev commented on validation-test/IDE/crashers_fixed/006-swift-declcontext-getprotocolself.swift:
```diff
@@ -1,3 +1,6 @@
+// We disable USR to Decl verification since the USR for declarations within the
+// extension with no type name have a USR that looks like a top-level declaration.
+
 // RUN: %target-swift-ide-test -code-completion -code-completion-token=A -source-filename=%s
```
Yeah, sorry I missed that 😅 Fixed.

hamishknight commented:
@swift-ci please test

a7medev commented:
@hamishknight It's going to fail again since we've neither addressed nor fixed the issues in tests that import Cocoa, I think we can cancel it until we decide on whether we're going to put time into fixing this or if we're gonna skip it for now. wdyt?

a7medev committed:
[Mangler] Mangle compatibility aliases as decls synthesized by ClangImporter

hamishknight commented on test/APINotes/versioned-test-mangling.swift:
```diff
@@ -8,7 +8,7 @@
 
-// CHECK-TOP-ALIAS-4: typealias ImportantCStruct = VeryImportantCStruct
+// CHECK-TOP-ALIAS-4: struct VeryImportantCStruct {
 // CHECK-TOP-ALIAS-5: struct VeryImportantCStruct {
-// CHECK-NESTED-ALIAS-4: typealias InnerInSwift5 = Outer.Inner
+// CHECK-NESTED-ALIAS-4: struct Inner {
 // CHECK-NESTED-ALIAS-5: struct Inner {
```
Can you also add checks for the typealias using the new compatibility alias mangling?

a7medev replied:
Added. Please recheck 🙏🏼

hamishknight commented on lib/AST/ASTMangler.cpp:
```diff
@@ -3116,6 +3124,13 @@ void ASTMangler::appendAnyGenericType(const GenericTypeDecl *decl,
   // Always use Clang names for imported Clang declarations, unless they don't
   // have one.
   auto tryAppendClangName = [this, decl]() -> bool {
+    // We use the Swift name rather than the Clang name for compatibility
+    // aliases to avoid multiple declarations having the same mangled name.
```
More precisely it's because `ClangImporterContext` entities are always mangled according to their Swift name, since that's what the demangler expects

hamishknight reviewed (commented):
Thanks!

a7medev committed:
[Test] Add checks for typealias mangling as Clang-importer-synthesized

hamishknight reviewed (approved):
Let me double check with the Clang Importer folks that they're happy with the compatibility alias change, but this LGTM thank you!

a7medev commented:
@hamishknight They're failing now. There seems to be an issue in the shell script it's trying to run (an unmatched single quote).

```
/Users/ec2-user/jenkins/workspace/swift-PR-macos@tmp/durable-de9ff1d2/script.sh.copy: line 1: unexpected EOF while looking for matching `''
```

a7medev committed:
[IDE] Provide raw & XML documentation for completion items

a7medev committed:
[Test] Enable USR to Decl verification in IDE/complete_crashes & IDE/complete_where_clause

a7medev committed:
[Test] Add sort by name option to swift-ide-test for completion

a7medev committed:
[Test] Update completion doc tests to check for raw doc comment

a7medev commented on test/IDE/complete_crashes.swift:
```diff
@@ -194,8 +194,14 @@ struct S_RDAR_28991372 {
+//        triggering module lookup during the first pass of IDE inspection
+//        (when not all decls have been parsed yet) and caching the top-level
+//        decls in the module's SourceCacheLookup ignoring decls added later
+//        through SourceFile::addTopLevelDecl.
```
This issue has been fixed in #83613 so no need to add these.

a7medev commented:
@hamishknight @rintaro I've updated the PR to disable USR to declaration verification in `IDE/crashers_fixed/036-swift-constraints-constraintsystem-matchdeepequalitytypes.swift` due to `extension{`.

I've also rebased on `main` instead of merging which I used previously.

Can you please re-trigger CI? 🙏🏼

a7medev commented:
@hamishknight Sorry, I forgot to remove `CodeCompletionResultBuilder::addChunkWithText` when resolving the conflict between main and my branch.

Can you please re-trigger tests? 🙏🏼

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test
